### PR TITLE
[SPARK-32874][SQL][TEST] Enhance result set meta data check for execute statement operation with thrift server

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR adds test cases for the result set metadata checking for Spark's `ExecuteStatementOperation` to make the JDBC API more future-proofing because any server-side change may affect the client compatibility.

### Why are the changes needed?

add test to prevent potential silent behavior change for JDBC users.

### Does this PR introduce _any_ user-facing change?

NO, test only 
### How was this patch tested?

add new test

### CODE REVIEW

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
cc @cloud-fan @wangyum @juliuszsompolski @maropu thanks very much~

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala [hunk 1]
```diff
@@ -148,49 +148,88 @@ class SparkThriftServerProtocolVersionsSuite extends HiveThriftJdbcTest {
       testExecuteStatementWithProtocolVersion(version, "SELECT cast(1 as byte)") { rs =>
         assert(rs.next())
         assert(rs.getByte(1) === 1.toByte)
+        val metaData = rs.getMetaData
+        assert(metaData.getColumnName(1) === "CAST(1 AS TINYINT)")
+        assert(metaData.getColumnType(1) === java.sql.Types.TINYINT)
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:153]: We don't need to check `getColumnTypeName`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala [hunk 2]
```diff
@@ -148,49 +148,88 @@ class SparkThriftServerProtocolVersionsSuite extends HiveThriftJdbcTest {
...
+        assert(metaData.getColumnType(1) === java.sql.Types.DOUBLE)
+        assert(metaData.getPrecision(1) === 15)
+        assert(metaData.getScale(1) === 15)
       }
     }
 
     test(s"$version get decimal type") {
       testExecuteStatementWithProtocolVersion(version,
-        "SELECT cast(1 as decimal(18, 2)) as c") { rs =>
+        "SELECT cast(1 as decimal(9, 1)) as col0, 1234.56BD as col1, 0.123 as col2") { rs =>
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:221]: Why did you modify the existing statement?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala [hunk 3]
```diff
@@ -148,49 +148,88 @@ class SparkThriftServerProtocolVersionsSuite extends HiveThriftJdbcTest {
...
         assert(rs.next())
-        assert(rs.getBigDecimal(1) === new java.math.BigDecimal("1.00"))
+        assert(rs.getBigDecimal(1) === new java.math.BigDecimal("1.0"))
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:223]: We don't need to test `rs.getBigDecimal(2)`?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:223]: It's as same as L224

**Comment by cloud-fan** [comment]
cc @bogdanghit

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:221]: for easily building the following  check loop

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala:154]: thanks for the suggestions~

**Review by juliuszsompolski — APPROVED** [review]
LGTM.
cc @bogdanghit

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by maropu** [comment]
@yaooqinn Looks this commit breaks`spark-master-test-sbt-hadoop-2.7-hive-1.2`, so could you check it? (We forgot to test this with the tag `[test-hive1.2][test-hadoop2.7]`...)
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7-hive-1.2/

**Comment by yaooqinn** [comment]
oh my... I will fix this in a follow-up as it looks pretty clear in the test failure.

BTW, do we have a plan to remove hive-1.2 support from compile? 🥺

**Comment by maropu** [comment]
> BTW, do we have a plan to remove hive-1.2 support from compile? 🥺

At least, I don't see any discussion about it.

**Pull request closed without merging**